### PR TITLE
Allow "cycles" through test suite dependencies

### PR DESCRIPTION
**Pull request by 23Skidoo**
A version of #3232/#3290 updated to work with `master`. Also moved the branch to the upstream repo so that others can push to it.

/cc @ezyang @edsko

### CODE REVIEW

**Comment by 23Skidoo** [comment]
Current status:

1. Needs a review from @kosmikus.
1.  There's a bug in interaction with sandboxes, see https://github.com/haskell/cabal/pull/3290#issuecomment-206511922. I won't have time to look into this for a while, so help appreciated.
1. There's also a failing test:

```
      indepGoals2:                                      FAIL
        expected: Just [("A",1),("B",1),("C",1),("D",1)]
         but got: Just [("A",1),("B",1),("C",1),("D",1),("D",2)]
```

**Comment by 23Skidoo** [comment]
I guess this also warrants a note in the changelog.

**Comment by 23Skidoo** [comment]
Hmm, apparently we have a pre-receive hook that doesn't allow force-pushing to all branches (not just protected ones).

**Comment by hvr** [comment]
I just tried this against `time-1.6.0.1` from Git w/ GHC 7.10... but I'm not sure this works as expected: the in-place lib:time-1.6.0.1 seems to be only used by time's testsuite, but not by its dependencies such as `random`/`tf-random` etc... 

if I interpret the `nm` output right, `tests` is in fact linked against two versions of `time`:

```
$ nm -g ./dist-newstyle/build/time-1.6.0.1/build/tests/tests | grep DataziTimeziCalendarziWeekDate_fromWeekDate2
0000000000a5d0c8 D timezuFTheb6LSxyX1UABIbBXRfn_DataziTimeziCalendarziWeekDate_fromWeekDate2_closure
0000000000a30028 D z34UAg7ZZnVmmgR6Tt8tOJ3oLE_DataziTimeziCalendarziWeekDate_fromWeekDate2_closure
```

/cc @ezyang

**Comment by kosmikus** [comment]
I'm not completely sure why this is perceived to be blocked by me. There seem to be more non-solver changes than solver changes here, and there was some discussion that the patch is still buggy, which at least according to @edsko is unlikely to be due to the solver part.

For me to contribute further on this, I'd need someone who knows the history of this patch and understands the changes that are made to the non-solver part to explain to me what the issue is and what we are supposed to achieve here.

**Comment by 23Skidoo** [comment]
@kosmikus Can you take a look at the solver parts of the patch at least? IIUC, the solver side is considered complete.

**Comment by 23Skidoo** [comment]
@kosmikus 

> For me to contribute further on this, I'd need someone who knows the history of this patch and understands the changes that are made to the non-solver part to explain to me what the issue is and what we are supposed to achieve here.

The rationale for this change is given in #3232. @edsko is the one to talk to about the implementation details.

**Comment by kosmikus** [comment]
Ok, perhaps I was confused then. I thought others had worked on this as well, and that it had something to do with sandboxes.

**Comment by hvr** [comment]
Now I see that in #3232 it's mentioned that linking against multiple library versions is by design

File: cabal-install/Distribution/Solver/Modular/Dependency.hs [hunk 1]
```diff
@@ -212,8 +219,13 @@ data QualifyOptions = QO {
 --
 -- NOTE: It's the _dependencies_ of a package that may or may not be independent
 -- from the package itself. Package flag choices must of course be consistent.
+--
+-- NOTE 2: This should be called on _all_ dependencies of a package. If it gets
+-- called on a subset of the dependencies, we might construct invalid
+-- quantifiers. In particular, we might conclude that a dependency of a test
+-- suite is not shared with the library and hence is independent.
```
Comment (grayjay) [inline cabal-install/Distribution/Solver/Modular/Dependency.hs]: I think this is inconsistent with #3357.  It added `requalifyDeps`, which can call `qualifyDeps` on only a subset of dependencies: https://github.com/haskell/cabal/blob/37325ca24b37ab2a96b1947b6da301c5d012f168/cabal-install/Distribution/Solver/Modular/Linking.hs#L290

**Comment by grayjay** [comment]
Should this feature be controlled by a flag, since most test suites don't have cyclic dependencies, and inconsistent dependencies can lead to compile errors?

**Comment by edsko** [comment]
@grayjay's comment that

> I think this is inconsistent with #3357. It added requalifyDeps, which can call qualifyDeps on only a subset of dependencies: https://github.com/haskell/cabal/blob/37325ca24b37ab2a96b1947b6da301c5d012f168/cabal-install/Distribution/Solver/Modular/Linking.hs#L290

is (as per usual.. :sweat: ) an excellent observation and something that needs fixing before this can be merged. Sadly the fix won't be entirely trivial.

**Comment by dcoutts** [comment]
> Should this feature be controlled by a flag, since most test suites don't have cyclic dependencies, and inconsistent dependencies can lead to compile errors?

@grayjay the hope is that this can actually be fully automatic. What kind of cases can lead to compile errors?

Here's an example: suppose `optparse-appicative` has a testsuite that uses `tasty` and suppose that this testsuite wants to makes use of `tasty`'s feature to add new command line arguments which itself makes use of `optparse-appicative`. Does this testsuite really expect that the `tasty` lib is built with the instance of `optparse-appicative` that is under test? If it does, then yes it could do that, but I would say that this is an unreasonable expectation.

Or another example might be a testsuite or benchmark for bytestring or binary. If these testsuites/benchmarks in their support code need to call functions from their testing/benchmarking libs (QC, criterion, tasty etc) that themselves use e.g. the `ByteString` type or the `Binary` class in their interface then yes that becomes tricky (but not impossible) since they cannot use any function from the bytestring lib directly since the visible one is the one under test.

But again I don't think this is a common problem or a reasonable expectation that the support libs for your testsuite or benchmark code are themselves compiled against the code under test. Indeed I think the reverse expectation is much more reasonable. If I'm testing bytestring or binary or something, then I would much prefer it if criterion/quickcheck/tasty etc were built against existing stable versions of bytestring, binary etc than to try and build them against the code I'm currently testing. Also, this avoids having to rebuild all those support libs every time I make a minor change to the code under test.

Another argument is that while yes in some cases there would be limitations on what you can do in the testsuite/benchmark since it's hard to make use of two versions of the same package (though we may be able to do something about this with package-qualified imports), these are not cases that work at all right now. So we're allowing more.

If you or anyone can spot cases where it's not as rosy as this, please do say so.

**Comment by 23Skidoo** [comment]
This branch needed rebasing, so I deleted and recreated it (couldn't force-push due to the pre-receive hook complaining). But apparently I now can't reopen this PR, so I'll open another.

**Comment by 23Skidoo** [comment]
Opened #3422.

@hvr, is there some way to get rid of the pre-receive hook? I think that only `master` and release branches should be protected (and they already are), for other branches we should allow force-pushing.

**Comment by 23Skidoo** [comment]
Asked support to re-enable force-pushes.

**Comment by 23Skidoo** [comment]
Force-pushes now work again on feature branches.

**Comment by hvr** [comment]
@23Skidoo for the record;  the global force-push-protection predates the "protected branches" feature; in the past you had to ask the github admins to enable it repo-wide. Luckily now there's more granular protected branches (even though I'm still not satisfied by those, but it's better than nothing :-) )

**Comment by 23Skidoo** [comment]
> for the record; the global force-push-protection predates the "protected branches" feature;

Yeah, I figured as much.

**Pull request closed without merging**